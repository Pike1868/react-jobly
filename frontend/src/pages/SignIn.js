import React, { useState } from "react";
import {
  Avatar,
  Box,
  Button,
  Container,
  Grid,
  Link,
  Typography,
  TextField,
} from "@mui/material";
import LockOutlinedIcon from "@mui/icons-material/LockOutlined";
import JoblyApi from "../api/JoblyApi";
import { useUserContext } from "../context/UserContext";
import { jwtDecode } from "jwt-decode";

export default function SignIn() {
  const { setUser } = useUserContext();
  const [error, setError] = useState("");

  const handleSubmit = async (event) => {
    event.preventDefault();
    setError(null);
    const data = new FormData(event.currentTarget);
    const username = data.get("username");
    const password = data.get("password");

    try {
      const token = await JoblyApi.login(username, password);
      const decoded = jwtDecode(token);
      if (decoded) {
        setUser({ username: decoded.username, isAdmin: decoded.isAdmin });
      }
    } catch (err) {
      setError(err || "An error occurred during sign in, please try again.");
    }
  };

  return (
    <Container component="main" maxWidth="xs">
      <Box
        sx={{
          marginTop: 8,
          display: "flex",
          flexDirection: "column",
          alignItems: "center",
        }}
      >
        <Avatar sx={{ m: 1, bgcolor: "#253031", p: 2 }}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h5">
          Sign in
        </Typography>

        {error && <Typography color="error">{error}</Typography>}

        <Box component="form" onSubmit={handleSubmit} sx={{ mt: 1 }}>
          <TextField
            margin="normal"
            required
            fullWidth
            id="username"
            label="Username"
            name="username"
            autoComplete="username"
            autoFocus
          />
          <TextField
            margin="normal"
            required
            fullWidth
            name="password"
            label="Password"
            type="password"
            id="password"
            autoComplete="current-password"
          />
          <Button
            type="submit"
            fullWidth
            variant="contained"
            sx={{ mt: 3, mb: 2 }}
          >
            Sign In
          </Button>
          <Grid container>
            <Grid item>
              <Link href="/signup" variant="body2">
                Don't have an account? Sign Up!
              </Link>
            </Grid>
          </Grid>
        </Box>
      </Box>
    </Container>
  );
}
