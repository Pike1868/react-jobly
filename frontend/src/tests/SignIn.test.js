import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import SignIn from "../pages/SignIn";
import { mockUserDetails } from "./commonSetup";

jest.mock("../context/UserContext", () => ({
  useUserContext: () => mockUserDetails,
}));

describe("SignIn", () => {
  it("renders without crashing", () => {
    render(
      <MemoryRouter>
        <SignIn />
      </MemoryRouter>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <SignIn />
      </MemoryRouter>
    );

    expect(asFragment).toMatchSnapshot();
  });
});
