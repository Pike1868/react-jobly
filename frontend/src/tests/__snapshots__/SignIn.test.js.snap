// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn matches snapshot 1`] = `[Function]`;
